File: backend/gateway/src/auth-rest/internal-service-auth.service.ts
import {
  Injectable,
  InternalServerErrorException,
  Logger,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { HttpService } from '@nestjs/axios';
import { firstValueFrom } from 'rxjs';

type ServiceTokenResponse = {
  data?: {
    accessToken?: string;
  };
};

@Injectable()
export class InternalServiceAuthService {
  private readonly logger = new Logger(
    InternalServiceAuthService.name,
  );
  private readonly ms4RestUrl: string;
  private readonly clientId: string;
  private readonly clientSecret: string;
  private readonly cacheSkewMs: number;

  private cachedToken: string | null = null;
  private cachedTokenExpiresAt = 0;
  private pendingTokenRequest:
    | Promise<string>
    | null = null;

  constructor(
    private readonly configService: ConfigService,
    private readonly httpService: HttpService,
  ) {
    const ms4RestUrl =
      this.configService.get<string>('ms4.restUrl');

    if (!ms4RestUrl) {
      throw new InternalServerErrorException(
        'MS4_REST_URL is not configured',
      );
    }

    this.ms4RestUrl = ms4RestUrl;
    this.clientId =
      this.configService.get<string>(
        'internalAuth.clientId',
      ) || '';
    this.clientSecret =
      this.configService.get<string>(
        'internalAuth.clientSecret',
      ) || '';
    this.cacheSkewMs =
      this.configService.get<number>(
        'internalAuth.cacheSkewMs',
      ) || 60000;
  }

  async getAuthorizationHeader(): Promise<string> {
    const accessToken =
      await this.getAccessToken();
    return `Bearer ${accessToken}`;
  }

  async getAccessToken(): Promise<string> {
    const now = Date.now();
    if (
      this.cachedToken &&
      now <
        this.cachedTokenExpiresAt -
          this.cacheSkewMs
    ) {
      return this.cachedToken;
    }

    if (this.pendingTokenRequest) {
      return this.pendingTokenRequest;
    }

    this.pendingTokenRequest =
      this.fetchAccessToken();

    try {
      return await this.pendingTokenRequest;
    } finally {
      this.pendingTokenRequest = null;
    }
  }

  private async fetchAccessToken(): Promise<string> {
    if (!this.clientId || !this.clientSecret) {
      throw new InternalServerErrorException(
        'Las credenciales internas no estan configuradas en el gateway.',
      );
    }

    const response =
      await firstValueFrom(
        this.httpService.post<ServiceTokenResponse>(
          `${this.ms4RestUrl}/auth/service-token`,
          {
            clientId: this.clientId,
            clientSecret: this.clientSecret,
          },
        ),
      );

    const accessToken =
      response.data?.data?.accessToken;

    if (!accessToken) {
      throw new InternalServerErrorException(
        'El servicio auth no devolvio un accessToken tecnico valido.',
      );
    }

    this.cachedToken = accessToken;
    this.cachedTokenExpiresAt =
      this.extractExpirationTime(
        accessToken,
      );

    this.logger.debug(
      `Token tecnico actualizado para ${this.clientId}.`,
    );

    return accessToken;
  }

  private extractExpirationTime(
    token: string,
  ): number {
    const parts = token.split('.');
    if (parts.length !== 3) {
      return Date.now();
    }

    try {
      const payload = JSON.parse(
        Buffer.from(
          parts[1],
          'base64url',
        ).toString('utf8'),
      ) as { exp?: number };

      if (
        typeof payload.exp === 'number'
      ) {
        return payload.exp * 1000;
      }
    } catch (error) {
      this.logger.warn(
        `No se pudo leer la expiracion del token tecnico: ${error instanceof Error ? error.message : 'error desconocido'}`,
      );
    }

    return Date.now();
  }
}

File: backend/ms-restaurante/src/main/java/com/restaurante/services/InternalServiceAuthService.java
package com.restaurante.services;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Base64;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

@Service
public class InternalServiceAuthService {

    private static final Logger logger = LoggerFactory.getLogger(InternalServiceAuthService.class);
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private final RestClient restClient;
    private final String authBaseUrl;
    private final String clientId;
    private final String clientSecret;
    private final long cacheSkewSeconds;

    private String cachedAccessToken;
    private Instant cachedExpiresAt;

    public InternalServiceAuthService(
            RestClient.Builder restClientBuilder,
            @Value("${app.internal-auth.base-url}") String authBaseUrl,
            @Value("${app.internal-auth.client-id}") String clientId,
            @Value("${app.internal-auth.client-secret:}") String clientSecret,
            @Value("${app.internal-auth.cache-skew-seconds:60}") long cacheSkewSeconds
    ) {
        this.restClient = restClientBuilder.build();
        this.authBaseUrl = authBaseUrl;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.cacheSkewSeconds = cacheSkewSeconds;
    }

    public synchronized String getAccessToken() {
        if (cachedAccessToken != null
                && cachedExpiresAt != null
                && Instant.now().isBefore(cachedExpiresAt.minusSeconds(cacheSkewSeconds))) {
            return cachedAccessToken;
        }

        if (clientSecret == null || clientSecret.isBlank()) {
            throw new IllegalStateException(
                    "La credencial interna no esta configurada en app.internal-auth.client-secret"
            );
        }

        JsonNode responseBody = restClient.post()
                .uri(authBaseUrl + "/auth/service-token")
                .contentType(MediaType.APPLICATION_JSON)
                .body(new ServiceTokenRequest(clientId, clientSecret))
                .retrieve()
                .body(JsonNode.class);

        String accessToken = extractAccessToken(responseBody);
        cachedAccessToken = accessToken;
        cachedExpiresAt = extractExpiration(accessToken);

        logger.debug("Token tecnico actualizado para {}", clientId);
        return accessToken;
    }

    public String getAuthorizationHeader() {
        return "Bearer " + getAccessToken();
    }

    private String extractAccessToken(JsonNode responseBody) {
        JsonNode accessTokenNode = responseBody != null
                ? responseBody.path("data").path("accessToken")
                : null;

        if (accessTokenNode == null || accessTokenNode.isMissingNode() || accessTokenNode.asText().isBlank()) {
            throw new IllegalStateException("Auth no devolvio un accessToken tecnico valido");
        }

        return accessTokenNode.asText();
    }

    private Instant extractExpiration(String accessToken) {
        try {
            String[] parts = accessToken.split("\\.");
            if (parts.length != 3) {
                return Instant.now();
            }

            String payloadJson = new String(
                    Base64.getUrlDecoder().decode(padBase64(parts[1])),
                    StandardCharsets.UTF_8
            );
            JsonNode payload = OBJECT_MAPPER.readTree(payloadJson);
            long exp = payload.path("exp").asLong(0L);

            if (exp > 0L) {
                return Instant.ofEpochSecond(exp);
            }
        } catch (Exception error) {
            logger.warn("No se pudo leer la expiracion del token tecnico: {}", error.getMessage());
        }

        return Instant.now();
    }

    private String padBase64(String value) {
        int remainder = value.length() % 4;
        if (remainder == 0) {
            return value;
        }
        return value + "=".repeat(4 - remainder);
    }

    private record ServiceTokenRequest(
            String clientId,
            String clientSecret
    ) {
    }
}
